File: game/src/main/java/com/bartlomiejpluta/base/game/map/serial/ProtobufMapDeserializer.java
package com.bartlomiejpluta.base.game.map.serial;

import com.bartlomiejpluta.base.core.error.AppException;
import com.bartlomiejpluta.base.core.util.mesh.MeshManager;
import com.bartlomiejpluta.base.game.image.manager.ImageManager;
import com.bartlomiejpluta.base.game.map.layer.image.ImageLayerMode;
import com.bartlomiejpluta.base.game.map.layer.object.PassageAbility;
import com.bartlomiejpluta.base.game.map.model.GameMap;
import com.bartlomiejpluta.base.game.tileset.manager.TileSetManager;
import com.bartlomiejpluta.base.proto.GameMapProto;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.InputStream;

@Slf4j
@Component
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class ProtobufMapDeserializer extends MapDeserializer {
   private final MeshManager meshManager;
   private final TileSetManager tileSetManager;
   private final ImageManager imageManager;

   @Override
   protected GameMap parse(InputStream input) throws Exception {
      var proto = GameMapProto.GameMap.parseFrom(input);
      var tileSet = tileSetManager.loadTileSet(proto.getTileSetUID());
      var map = new GameMap(tileSet, proto.getRows(), proto.getColumns(), proto.getHandler());

      proto.getLayersList().forEach(layer -> deserializeLayer(map, layer));

      return map;
   }

   private void deserializeLayer(GameMap map, GameMapProto.Layer proto) {
      if (proto.hasTileLayer()) {
         deserializeTileLayer(map, proto);
      } else if (proto.hasObjectLayer()) {
         deserializeObjectLayer(map, proto);
      } else if (proto.hasColorLayer()) {
         deserializeColorLayer(map, proto);
      } else if (proto.hasImageLayer()) {
         deserializeImageLayer(map, proto);
      } else {
         throw new AppException("Not supported layer type");
      }
   }

   private void deserializeTileLayer(GameMap map, GameMapProto.Layer proto) {
      var layer = map.createTileLayer();
      var columns = map.getColumns();
      var tiles = proto.getTileLayer().getTilesList();

      for (var i = 0; i < tiles.size(); ++i) {
         var tile = tiles.get(i);

         if (tile == 0) {
            map.clearTile(layer, i / columns, i % columns);
         } else {
            map.setTile(layer, i / columns, i % columns, tile - 1);
         }
      }
   }

   private void deserializeObjectLayer(GameMap map, GameMapProto.Layer proto) {
      var layer = map.createObjectLayer();
      var columns = map.getColumns();
      var passageMap = proto.getObjectLayer().getPassageMapList();

      for (var i = 0; i < passageMap.size(); ++i) {
         map.setPassageAbility(layer, i / columns, i % columns, switch (passageMap.get(i)) {
            case ALLOW -> PassageAbility.ALLOW;
            case BLOCK -> PassageAbility.BLOCK;
            case RIGHT_ONLY -> PassageAbility.RIGHT_ONLY;
            case LEFT_ONLY -> PassageAbility.LEFT_ONLY;
            case DOWN_ONLY -> PassageAbility.DOWN_ONLY;
            case UP_ONLY -> PassageAbility.UP_ONLY;
         });
      }
   }

   private void deserializeColorLayer(GameMap map, GameMapProto.Layer proto) {
      var protoColorLayer = proto.getColorLayer();
      map.createColorLayer(
            meshManager,
            protoColorLayer.getRed() / 100.0f,
            protoColorLayer.getGreen() / 100.0f,
            protoColorLayer.getBlue() / 100.0f,
            protoColorLayer.getAlpha() / 100.0f
      );
   }

   private void deserializeImageLayer(GameMap map, GameMapProto.Layer proto) {
      var protoImageLayer = proto.getImageLayer();
      var image = imageManager.loadImage(protoImageLayer.getImageUID());

      var mode = switch (proto.getImageLayer().getMode()) {
         case NORMAL -> ImageLayerMode.NORMAL;
         case FIT_MAP -> ImageLayerMode.FIT_MAP;
         case FIT_SCREEN -> ImageLayerMode.FIT_SCREEN;
      };

      map.createImageLayer(
            image,
            protoImageLayer.getOpacity() / 100.0f,
            protoImageLayer.getX(),
            protoImageLayer.getY(),
            (float) protoImageLayer.getScaleX(),
            (float) protoImageLayer.getScaleY(),
            mode,
            protoImageLayer.getParallax()
      );
   }
}

File: editor/src/main/kotlin/com/bartlomiejpluta/base/editor/map/serial/ProtobufMapDeserializer.kt
package com.bartlomiejpluta.base.editor.map.serial

import com.bartlomiejpluta.base.editor.map.model.enumeration.ImageLayerMode
import com.bartlomiejpluta.base.editor.map.model.enumeration.PassageAbility
import com.bartlomiejpluta.base.editor.map.model.layer.*
import com.bartlomiejpluta.base.editor.map.model.map.GameMap
import com.bartlomiejpluta.base.editor.map.model.obj.MapObject
import com.bartlomiejpluta.base.editor.project.context.ProjectContext
import com.bartlomiejpluta.base.editor.tileset.model.Tile
import com.bartlomiejpluta.base.proto.GameMapProto
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component
import java.io.InputStream

@Component
class ProtobufMapDeserializer : MapDeserializer {

   @Autowired
   private lateinit var appContext: ApplicationContext

   private val projectContext: ProjectContext by lazy {
      appContext.getBean(ProjectContext::class.java)
   }

   override fun deserialize(input: InputStream) = deserialize(input, { _, uid -> uid }, { _, uid -> uid })

   override fun deserialize(
      input: InputStream,
      replaceTileSet: (String, String) -> String,
      replaceAutoTile: (String, String) -> String
   ): GameMap {
      val proto = GameMapProto.GameMap.parseFrom(input)
      val map = GameMap(proto.tileWidth.toDouble(), proto.tileHeight.toDouble())
      map.uid = proto.uid
      map.rows = proto.rows
      map.columns = proto.columns
      map.handler = proto.handler

      proto.layersList
         .filter { it.hasTileLayer() || it.hasAutoTileLayer() || it.hasObjectLayer() || it.hasColorLayer() }
         .forEach { map.layers.add(deserializeLayer(map.rows, map.columns, it, replaceTileSet, replaceAutoTile)) }

      return map
   }

   private fun deserializeLayer(
      rows: Int,
      columns: Int,
      proto: GameMapProto.Layer,
      replaceTileSet: (String, String) -> String,
      replaceAutoTile: (String, String) -> String
   ): Layer {
      return when {
         proto.hasTileLayer() -> deserializeTileLayer(rows, columns, proto, replaceTileSet)
         proto.hasAutoTileLayer() -> deserializeAutoTileLayer(rows, columns, proto, replaceAutoTile)
         proto.hasObjectLayer() -> deserializeObjectLayer(rows, columns, proto)
         proto.hasColorLayer() -> deserializeColorLayer(proto)
         proto.hasImageLayer() -> deserializeImageLayer(proto)

         else -> throw IllegalStateException("Not supported layer type")
      }
   }

   private fun deserializeTileLayer(
      rows: Int,
      columns: Int,
      proto: GameMapProto.Layer,
      replaceTileSet: (String, String) -> String
   ): Layer {
      val layer: Array<Array<Tile?>> = Array(rows) { Array(columns) { null } }
      val tileSetAsset = projectContext.findTileSetAsset(replaceTileSet(proto.name, proto.tileLayer.tilesetUID))
      val tileSet = projectContext.loadTileSet(tileSetAsset.uid)

      proto.tileLayer.tilesList.forEachIndexed { index, tile ->
         layer[index / columns][index % columns] = when (tile) {
            0 -> null
            else -> tileSet.getTile(tile - 1)
         }
      }

      return TileLayer(proto.name, rows, columns, tileSetAsset, layer)
   }

   private fun deserializeAutoTileLayer(
      rows: Int,
      columns: Int,
      proto: GameMapProto.Layer,
      replaceTileSet: (String, String) -> String
   ): AutoTileLayer {
      val layer: Array<Array<Int>> = Array(rows) { Array(columns) { 0 } }
      val autoTile = projectContext.findAutoTileAsset(replaceTileSet(proto.name, proto.autoTileLayer.autotileUID))

      proto.autoTileLayer.tilesList.forEachIndexed { index, tile ->
         layer[index / columns][index % columns] = tile
      }

      return AutoTileLayer(
         proto.name,
         rows,
         columns,
         autoTile,
         proto.autoTileLayer.animated,
         proto.autoTileLayer.animationDuration,
         proto.autoTileLayer.connect,
         layer
      )
   }

   private fun deserializeObjectLayer(rows: Int, columns: Int, proto: GameMapProto.Layer): Layer {
      val passageMap: Array<Array<PassageAbility>> = Array(rows) { Array(columns) { PassageAbility.ALLOW } }

      proto.objectLayer.passageMapList.forEachIndexed { index, passage ->
         passageMap[index / columns][index % columns] = when (passage!!) {
            GameMapProto.PassageAbility.ALLOW -> PassageAbility.ALLOW
            GameMapProto.PassageAbility.BLOCK -> PassageAbility.BLOCK
         }
      }

      val objects = proto.objectLayer.objectsList.map {
         MapObject(it.x, it.y, it.code)
      }

      return ObjectLayer(proto.name, rows, columns, objects, passageMap)
   }

   private fun deserializeColorLayer(proto: GameMapProto.Layer): Layer {
      return ColorLayer(
         name = proto.name,
         red = proto.colorLayer.red,
         green = proto.colorLayer.green,
         blue = proto.colorLayer.blue,
         alpha = proto.colorLayer.alpha
      )
   }

   private fun deserializeImageLayer(proto: GameMapProto.Layer): Layer {
      return ImageLayer(
         name = proto.name,
         imageAsset = projectContext.findImageAsset(proto.imageLayer.imageUID),
         opacity = proto.imageLayer.opacity,
         x = proto.imageLayer.x,
         y = proto.imageLayer.y,
         scaleX = proto.imageLayer.scaleX,
         scaleY = proto.imageLayer.scaleY,
         mode = when (proto.imageLayer.mode!!) {
            GameMapProto.ImageLayerMode.NORMAL -> ImageLayerMode.NORMAL
            GameMapProto.ImageLayerMode.FIT_MAP -> ImageLayerMode.FIT_MAP
            GameMapProto.ImageLayerMode.FIT_SCREEN -> ImageLayerMode.FIT_SCREEN
         },
         parallax = proto.imageLayer.parallax
      )
   }
}
